### PR TITLE
Add nnn recipe

### PR DESCRIPTION
**Pull request by annagrram**
Requested in #4083.

This is a backport recipe and patchset for v8.2.1 (The current released version).

The master branch includes a builtin support for Haiku which, of course, will be available in next releases.

### CODE REVIEW

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+It runs smoothly on the Raspberry Pi, Termux on Android, Linux, macOS, BSD, Cygwin, WSL and works seamlessly with DEs and GUI utilities.
+
+Visit the Wiki for concepts, program usage, how-tos and troubleshooting.
+"
+HOMEPAGE="https://github.com/jarun/nnn"
+COPYRIGHT="2016-2019 Arun Prakash Jana"
+LICENSE="BSD (2-clause)"
+REVISION="1"
+SOURCE_URI="https://github.com/jarun/nnn/archive/v2.8.1.tar.gz"
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
SOURCE_URI="https://github.com/jarun/nnn/archive/v$portVersion.tar.gz"
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+PATCHES="nnn-$portVersion.patchset"
+
+ARCHITECTURES="x86 x86_64"
+
+PROVIDES="
+	nnn = $portVersion
+	cmd:nnn = $portVersion
+	"
+REQUIRES="
+        haiku
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
	haiku
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	lib:libncurses
+	lib:libreadline
+        "
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
	"
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+BUILD_REQUIRES="
+        haiku_devel
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
	haiku_devel
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	pkgconfig
+	devel:libncurses
+	devel:libreadline
+        "
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
	"
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+
+BUILD_PREREQUIRES="
+	cmd:make
+	cmd:gcc
+	cmd:g++
+	cmd:ld
+	cmd:install
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]: These should be alphabetically ordered

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+"
+HOMEPAGE="https://github.com/jarun/nnn"
+COPYRIGHT="2016-2019 Arun Prakash Jana"
+LICENSE="BSD (2-clause)"
+REVISION="1"
+SOURCE_URI="https://github.com/jarun/nnn/archive/v2.8.1.tar.gz"
+CHECKSUM_SHA256="cadf9cf8038433aeeb50a777180ad4b309ac7d2fec81c7da177ddca515812f06"
+PATCHES="nnn-$portVersion.patchset"
+
+ARCHITECTURES="x86 x86_64"
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]: x86 might be supported in SECONDARY_ARCHITECTURES.

```suggestion
ARCHITECTURES="x86_64"
```

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 8]
```diff
@@ -0,0 +1,54 @@
+SUMMARY="The missing terminal file manager for X"
+DESCRIPTION="
+nnn is a full-featured terminal file manager. It's tiny and nearly 0-config with an incredible performance.
+
+nnn is also a du analyzer, an app launcher, a batch renamer and a file picker. The plugin repository has tons of plugins and documentation to extend the capabilities further. You can plug new functionality and play with a custom keybind instantly. There's an independent (neo)vim plugin.
+
+It runs smoothly on the Raspberry Pi, Termux on Android, Linux, macOS, BSD, Cygwin, WSL and works seamlessly with DEs and GUI utilities.
+
+Visit the Wiki for concepts, program usage, how-tos and troubleshooting.
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]: Please wrap lines at 80-chars with \

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+	"
+REQUIRES="
+        haiku
+	lib:libncurses
+	lib:libreadline
+        "
+
+BUILD_REQUIRES="
+        haiku_devel
+	pkgconfig
```
Comment (korli) [inline app-misc/nnn/nnn-2.8.1.recipe]:
```suggestion
	cmd:pkg_config
```

Comment (korli) [reply app-misc/nnn/nnn-2.8.1.recipe]: cmd:pkg_config should be moved to BUILD_PREREQUISITES.

**Comment by annagrram** [comment]
@korli Thank you very much for your input.
I made the requested changes.

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 10]
```diff
@@ -0,0 +1,60 @@
+SUMMARY="The missing terminal file manager for X"
+DESCRIPTION="
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: Start description on the same line

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+nnn is also a du analyzer, an app launcher, a batch renamer and a file picker. \
+The plugin repository has tons of plugins and documentation to extend the \
+capabilities further. You can plug new functionality and play with a \
+custom keybind instantly. There's an independent (neo)vim plugin.
+
+It runs smoothly on the Raspberry Pi, Termux on Android, Linux, macOS, BSD, \
+Cygwin, WSL and works seamlessly with DEs and GUI utilities.
+
+Visit the Wiki for concepts, program usage, how-tos and troubleshooting.
+"
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: Put closing double qoute at the end of previous line, no empty line in between

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+	lib:libncurses$secondaryArchSuffix
+	lib:libreadline$secondaryArchSuffix
+	"
+
+BUILD_REQUIRES="
+	haiku${secondaryArchSuffix}_devel
+	devel:libncurses$secondaryArchSuffix
+	devel:libreadline$secondaryArchSuffix
+	"
+
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: No empty line here

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	cmd:gcc$secondaryArchSuffix
+	cmd:install
+	cmd:ld$secondaryArchSuffix
+	cmd:make
+	cmd:pkg_config$secondaryArchSuffix
+	"
+
+BUILD()
+{
+	make
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: make $jobArgs

**Comment by Begasus** [comment]
Made some suggestions according to our guidelines: https://github.com/haikuports/haikuports/wiki/HaikuPorter-Guidelines

**Comment by annagrram** [comment]
@Begasus Thank you for the input.

I've made the changes noted but will look at the guidelines later today or tomorrow as I don't have time currently.

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+SOURCE_URI="https://github.com/jarun/nnn/archive/v$portVersion.tar.gz"
+CHECKSUM_SHA256="cadf9cf8038433aeeb50a777180ad4b309ac7d2fec81c7da177ddca515812f06"
+PATCHES="nnn-$portVersion.patchset"
+
+ARCHITECTURES="x86_64"
+SECONDARY_ARCHITECTURES="x86"
+
+PROVIDES="
+	nnn$secondaryArchSuffix = $portVersion
+	cmd:nnn$secondaryArchSuffix = $portVersion
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: Binary is installed in $prefix/bin, so:
```diff
-	cmd:nnn$secondaryArchSuffix = $portVersion
+	cmd:nnn = $portVersion
```

**Comment by Begasus** [comment]
> @Begasus Thank you for the input.
> 
> I've made the changes noted but will look at the guidelines later today or tomorrow as I don't have time currently.

NP :) just did a successful build for nnn_x86 :)

**Comment by Begasus** [comment]
Looks to be running OK, just need to figure out how it works :)

**Comment by annagrram** [comment]
> Looks to be running OK, just need to figure out how it works :)

If you mean how `nnn` works, I'll admit it requires some getting used to, but it is a slick and efficient FM, especially for those who like to do their file management without that rodent on the table :)

**Comment by Begasus** [comment]
Already found out how to open txt and image files, navigation is working also fine

**Comment by Begasus** [comment]
@korli the patch has been sent upstream afaik (@annagrram ?) Checked and works OK from what I can tell on x86_gcc2 (seconaryArchSuffix) and x86_64, think we're good to go :)

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 15]
```diff
@@ -0,0 +1,57 @@
...
+It runs smoothly on the Raspberry Pi, Termux on Android, Linux, macOS, BSD, \
+Cygwin, WSL and works seamlessly with DEs and GUI utilities.
+
+Visit the Wiki for concepts, program usage, how-tos and troubleshooting."
+HOMEPAGE="https://github.com/jarun/nnn"
+COPYRIGHT="2016-2019 Arun Prakash Jana"
+LICENSE="BSD (2-clause)"
+REVISION="1"
+SOURCE_URI="https://github.com/jarun/nnn/archive/v$portVersion.tar.gz"
+CHECKSUM_SHA256="cadf9cf8038433aeeb50a777180ad4b309ac7d2fec81c7da177ddca515812f06"
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: Suggestion downloaded file looks more familiar:
SOURCE_FILENAME="nnn-v$portVersion.tar.gz"

File: app-misc/nnn/nnn-2.8.1.recipe [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+	"
+
+BUILD()
+{
+	make $jobArgs
+}
+
+INSTALL()
+{
+	make PREFIX=$prefix MANPREFIX=$manDir install
```
Comment (Begasus) [inline app-misc/nnn/nnn-2.8.1.recipe]: Just checked out a srcGitRev from master there with your changes, added a comment on your merged PR @annagrram , the suggested change will remove the need to set MANPREFIX here :)

Comment (Begasus) [reply app-misc/nnn/nnn-2.8.1.recipe]: PS, isn't there a way to rename "misc/haiku/Makefile" to "Makefile.haiku" in their github ? (maybe overkill) :)

**Comment by annagrram** [comment]
The patch has been merged to master upstream, yes.

Your Makefile change suggestion sounds good. I'll take care of it.

Regarding the name, IMHO, I think it's an overkill because it is already placed under haiku directory. I thought of naming it haiku.mk at first, but dismissed the idea for the same reason.

**Comment by Begasus** [comment]
OK, no problem, would reduce the call for make (make -f misc/haiku/Makefile $jobArgs) :)

**Comment by Begasus** [comment]
Looks OK here :+1: 
nnn

**Comment by jarun** [comment]
> Already found out how to open txt and image files, navigation is working also fine

With the changes in https://github.com/jarun/nnn/pull/425 opening files should work without any issues.

Also, we have a [quickstart](https://github.com/jarun/nnn#quickstart). I'll add `open` from Haiku in the top row shortly.

**Comment by jarun** [comment]
@annagrram can you please add `file` as a package dependency? It's not installed by default on haiku. This will enable `NNN_USE_EDITOR`.

Or if there's another alternative on Haiku we may need to tweak the code.

**Comment by annagrram** [comment]
> @annagrram can you please add `file` as a package dependency? It's not installed by default on haiku. This will enable `NNN_USE_EDITOR`.
> 
> Or if there's another alternative on Haiku we may need to tweak the code.

Sure.
I'll do all the requested changes that I haven't done already later today.

**Comment by Begasus** [comment]
@annagrram @korli anything that needs to be done here? Otherwise it's good to merge

**Comment by annagrram** [comment]
> @annagrram @korli anything that needs to be done here? Otherwise it's good to merge

Adding `file` to the requirements.
Give me a minute, and I'll push it

**Comment by annagrram** [comment]
Done

**Comment by Begasus** [comment]
Checking _x86 to make sure :)

**Comment by Begasus** [comment]
> @annagrram can you please add `file` as a package dependency? It's not installed by default on haiku. This will enable `NNN_USE_EDITOR`.
> 
> Or if there's another alternative on Haiku we may need to tweak the code.

@jarun with "find" added to the recipe, I guess we have to wait until you update "open" for Haiku?
@annagrram https://github.com/jarun/nnn/commit/a36146f9c08dcaf759c087b3ed0998fdd95a88c7 would that be useful to also include in the current patch?

**Comment by Begasus** [comment]
Guess it does :) I could even launch Qupzilla by pressing Enter in nnn :+1: 
Could you add that patch too @annagrram :)

**Comment by annagrram** [comment]
> Guess it does :) I could even launch Qupzilla by pressing Enter in nnn +1
> Could you add that patch too @annagrram :)

Done.
Please give it a quick test to see that I haven't broken anything :) (I tested it, but I never trust myself)

**Comment by jarun** [comment]
> @jarun with "find" added to the recipe, I guess we have to wait until you update "open" for Haiku?

I think you mean `file`. Yes, it's required.

@annagrram please ensure `file` is added as a dep.

**Comment by annagrram** [comment]
> @annagrram please ensure file is added as a dep.

It is.

**Comment by Begasus** [comment]
Working OK! Thanks @annagrram and @jarun :+1:

**Comment by annagrram** [comment]
@jarun Once we stabilize this recipe and it gets merged, I'll update it upstream.

Apparently now ^_^

**Comment by jarun** [comment]
Thanks all!

**Comment by Begasus** [comment]
For future additions we'll wait for the next release, you can keep track there @annagrram ?

**Comment by annagrram** [comment]
Thank you very much @Begasus

**Comment by annagrram** [comment]
> For future additions we'll wait for the next release, you can keep track there @annagrram ?

Will do.

**Pull request merged**